### ISSUE TITLE
Command-buffer extension testing

### ISSUE THREAD

**Issue by EwanC** [KhronosGroup/OpenCL-CTS#1369]
In order for the [cl_khr_command_buffer extension](https://www.khronos.org/registry/OpenCL/specs/3.0-unified/html/OpenCL_Ext.html#cl_khr_command_buffer) to become fully ratified from its provisional status it needs CTS testing. This issue is a plan of areas that need to be covered to fully test the extension. By having this breakdown, we should be able to collaborate on the tasks needing to be done.

## Device Info Queries

Verify the following queries can be performed successfully and return
defined values:

- [x] `CL_DEVICE_COMMAND_BUFFER_CAPABILITIES_KHR`
- [x] `CL_DEVICE_COMMAND_BUFFER_REQUIRED_QUEUE_PROPERTIES_KHR`

EDIT: Done in test fixture

### 1.1 Individual command entry-points

For each command which can be recorded into a command-buffer, check that the
follow steps can be performed successfully:

* Create a new command-buffer targeting a single queue
* Record <command>
* Finalize
* Enqueue
* clFinish & verify result.

- [x] `clCommandBarrierWithWaitListKHR`
- [x] `clCommandCopyBufferKHR`
- [x] `clCommandCopyBufferRectKHR`
- [x] `clCommandCopyBufferToImageKHR`
- [x] `clCommandCopyImageKHR`
- [x] `clCommandCopyImageToBufferKHR`
- [x] `clCommandFillBufferKHR`
- [x] `clCommandFillImageKHR`
- [x] `clCommandNDRangeKernelKHR`

EDIT: Done in https://github.com/KhronosGroup/OpenCL-CTS/pull/1594

### 1.2 Interleave command-buffer enqueues with regular enqueues

> *Only* when the device supports optional Simultaneous Use

- [x] Test case for interleaving command-buffer enqueues with enqueues of core
      OpenCL commands:

* Create a new in-order command-buffer targeting a single queue
* Record `clCommandNDRangeKernel` command
* Finalize
* Enqueue command-buffer
* Enqueue an `clEnqueueNDRangeKernel` operation to the same queue
* Enqueue command-buffer again
* clFinish & verify through result that ordering of enqueues was preserved.

 **DONE**: Scenario tested by `InterleavedEnqueueTest` in https://github.com/KhronosGroup/OpenCL-CTS/pull/1368

#### 1.3.1 

- [x] substitution on queue without properties 

* Create a command-buffer recorded to a single queue with no (or required) properties
* Record `clCommandNDRangeKernel` command
* Finalize
* Enqueue to another command-queue created with the same properties.
* clFinish and verify result

EDIT: Added as `test_queue_substitution` in https://github.com/KhronosGroup/OpenCL-CTS/pull/1584

#### 1.3.2

- [x] substitution on queue with properties

* Create a command-buffer recorded to a single queue with additional properties (e.g profiling enabled) set
* Record `clCommandNDRangeKernel` command
* Finalize
* Enqueue to another command-queue created with the same properties.
* clFinish and verify result

EDIT: Added as `test_properties_queue_substitution` in https://github.com/KhronosGroup/OpenCL-CTS/pull/1584

#### 1.3.3

- [x] simultaneous use queue substitution 

> Requires Simulteaneous-use property

* Create a command-buffer recorded to a single queue with no (or required) properties
* Record `clCommandNDRangeKernel` command
* Finalize
* Enqueue command-buffer to another command-queue created with the same properties, blocked on a user event
* Enqueue command-buffer to the same command-queue as the previous enqueue, and blocked on the same user event
* set user event status to complete
* call clFinish on queues and verify results

EDIT:  Added as `test_simultaneous_queue_substitution` in https://github.com/KhronosGroup/OpenCL-CTS/pull/1584 after removing the restriction that the two command-buffer enqueues must be to different queues, as it's something more suitable for testing in cl_khr_command_buffer_mutable_dispatch.

### 1.4 `clGetCommandBufferInfoKHR`

Test that all the command-buffer info queries can be performed successfully:

- [x] `CL_COMMAND_BUFFER_NUM_QUEUES_KHR`
- [x] `CL_COMMAND_BUFFER_QUEUES_KHR`
- [x] `CL_COMMAND_BUFFER_REFERENCE_COUNT_KHR`
- [x] `CL_COMMAND_BUFFER_STATE_KHR`
- [x] `CL_COMMAND_BUFFER_PROPERTIES_ARRAY_KHR`

EDIT - Done in https://github.com/KhronosGroup/OpenCL-CTS/pull/1611

`CL_COMMAND_BUFFER_STATE_KHR` should be able to be reliably tested for `CL_COMMAND_BUFFER_STATE_RECORDING_KHR`, `CL_COMMAND_BUFFER_STATE_EXECUTABLE_KHR`, and `CL_COMMAND_BUFFER_STATE_PENDING_KHR `. However, `CL_COMMAND_BUFFER_STATE_INVALID_KHR ` is
more implementation dependent, so we don't need a test for that.

clRetainCommandBufferKHR and clReleaseCommandBufferKHR should be tested as part of `CL_COMMAND_BUFFER_REFERENCE_COUNT_KHR`

### 1.5 `clCreateCommandBufferKHR`

I think this entry-point will end up being tested implicitly as part of the other tests, so doesn't need bespoke tests.

### 1.6 Printf

- [x] Test cases for device side Printf

* check device support printf - CL_COMMAND_BUFFER_CAPABILITY_KERNEL_PRINTF_KHR 
* Record a command-buffer with containing kernel using printf
* Enqueue the command-buffer multiple times to the same queue, with clFinish in between, and check correct
  output was printed

EDIT: Done in https://github.com/KhronosGroup/OpenCL-CTS/pull/1592

#### 1.6.1 Simultaneous use printf

- [x] Test cases for device side printf with a simultaneous use command-buffer

> requires simultaneous use

* check device support printf - CL_COMMAND_BUFFER_CAPABILITY_KERNEL_PRINTF_KHR 
* Record a command-buffer with containing kernel using printf
* Create a user event
* Enqueue the command-buffer multiple times to the same queue blocked on the user event.
* Set the user event status to complete and call clFinish on the queue.
* verify correct output was printed

EDIT: done in https://github.com/KhronosGroup/OpenCL-CTS/pull/1592

### 1.7 clSetKernelArg

- [x] Test interactions of clSetKernelArg with command-buffers, i.e that it doesn't affect already recorded commands

* Record a command-buffer with kernel execution command
* change arguments with clSetKernelArg and record another kernel execution command
* finalize and enqueue command-buffer, then verify from output that clSetKernelArg didn't affect the first command

#### 1.7.1  clSetKernelArg  on pending command buffer

- [x] Test interactions of clSetKernelArg on a command-buffer pending execution

* Record a command-buffer with kernel execution command
* finalize and enqueue command-buffer a queue blocked on a user event.
* If simultaneous use is supported, enqueue the same command-buffer so that multiple time blocked on the same user event.
* change the kernel arguments with clSetKernelArg, which should have no effect.
* set user event status to complete and flush the queue.
* verify from results that clSetKernelArg had no effect.

EDIT: Done in https://github.com/KhronosGroup/OpenCL-CTS/pull/1593

## 2 Out-of-order command-queue

- [x] Test case for out-of-order command-buffer:
**DONE**: `OutOfOrderTest` added in https://github.com/KhronosGroup/OpenCL-CTS/pull/1368

* Query that the device supports out-of-order command-queues
* Record multiple commands using `cl_sync_point_khr` to maintain an ordering
  which can be tested
* Finalize
* Enqueue
* Flush and verify results.

### 2.1 Out-of-order with simultaneous use

- [x] Test an out-of-order command-buffer with simultaneous use

Perform the above test but enqueue the command-buffer multiple times blocked on a user event, set the user event to complete, call clFinish, and verify results.

EDIT - Done in https://github.com/KhronosGroup/OpenCL-CTS/pull/1596

## 3 Events

Test interaction of cl_events with command-buffers

### 3.1 Event info

Test `clGetEventInfo()` info queries from event returned by
command-buffer enqueue for the single command-queue to command-buffer case:

- [x] `CL_EVENT_COMMAND_TYPE` is `CL_COMMAND_COMMAND_BUFFER_KHR`
- [x] `CL_EVENT_COMMAND_QUEUE`
- [x] `CL_EVENT_CONTEXT`
- [x] `CL_EVENT_COMMAND_EXECUTION_STATUS`
- [x] `CL_EVENT_REFERENCE_COUNT`

Edit: Done in https://github.com/KhronosGroup/OpenCL-CTS/pull/1602

### 3.2 Command-buffer profiling

- [x] Test profiling of a command-buffer where all commands are recorded to a
      single command-queue:

* Create a command-buffer from a command-queue created with the
  `CL_QUEUE_PROFILING_ENABLE` property set
* Record a NDRange command, finalize and enqueue returning an event
* Check event profiling queries are valid

EDIT: Done in https://github.com/KhronosGroup/OpenCL-CTS/pull/1598

### 3.2 Command-buffer profiling with simultaneous use

> Requires device simultaneous use support

- [x] Test profiling a command-buffer with simultaneous use

* Create a simultaneous use command-buffer from a command-queue created with the
  `CL_QUEUE_PROFILING_ENABLE` property set
* Record a NDRange command, finalize.
* Create a user event
* enqueue command-buffer multilple times, returning an event, but blocking on the user event
* set user event to complete and clFinish
* check profiling queries from the events returned by command-buffer enqueues are valid

EDIT: Done in https://github.com/KhronosGroup/OpenCL-CTS/pull/1598

#### 3.3.1  Returned events

- [x] Test that an event returned by a command-buffer enqueue can be waited on by
      regular commands.
- [x] Test that an event returned by a command-buffer enqueue can be waited on an
      enqueue of the same command-buffer. 
- [x] Tests that a command buffer enqueue can wait on the enqueue of a different command buffer.
- [x] Test `clSetEventCallback` works correctly on an event returned by `clEnqueueCommandBufferKHR`.
- [x] Test `clWaitForEvents` on a single event returned from a `clEnqueueCommandBufferKHR`.
- [x] Test `clWaitForEvents`  on multiple events returned from different `clEnqueueCommandBufferKHR` calls.
- [x] Test command-buffer enqueue with event from a regular command from the same queue in wait-list.
- [x] Test command-buffer enqueue with an event returned by another queue in the wait-list.

#### 3.3.2 User events

- [x] Test command-buffer enqueue blocked by a single user-event in wait-list
**DONE**: `UserEventTest` added in https://github.com/KhronosGroup/OpenCL-CTS/pull/1368
- [x] Test command-buffer enqueue blocked by a multiple user-events in wait-list
- [x] Test `clSetEventCallback` works correctly on an user defined event waited by `clEnqueueCommandBufferKHR`

EDIT: 3.3 tests done in https://github.com/KhronosGroup/OpenCL-CTS/pull/1606

**Comment by EwanC** [comment]
MR adding basic testing we can use as a starting point https://github.com/KhronosGroup/OpenCL-CTS/pull/1368

**Comment by aharon-abramson** [comment]
1.2: You say: "Only when the device supports optional Simultaneous Use". However, the spec states that "command-buffers are repeatedly enqueued without blocking in user code" - this means that the simultaneity relates to the command buffers themselves.
3.3.1: "Test that an event returned by a command-buffer enqueue can be waited on an enqueue of the same command-buffer" - this does require simultaneous use.
I suggest you add a test where you query CL_DEVICE_COMMAND_BUFFER_REQUIRED_QUEUE_PROPERTIES_KHR and then create a queue with such properties for testing command buffers (since Mobieye's implementation requires CL_QUEUE_NO_SYNC_OPERATIONS_INTEL, without this all the tests would be skipped).

**Comment by EwanC** [comment]
Good points, I'll elaborate on these for the benefit of Mobica

> 1.2: You say: "Only when the device supports optional Simultaneous Use". However, the spec states that "command-buffers  are repeatedly enqueued without blocking in user code" - this means that the simultaneity relates to the command buffers themselves.

Simultaneous use allows a command-buffer to be submitted for execution again before a previous instance of the same command-buffer has completed. And if coming from a background familiar with Vulkan, it is equivalent to `VK_COMMAND_BUFFER_USAGE_SIMULTANEOUS_USE_BIT`

`CL_COMMAND_BUFFER_CAPABILITY_SIMULTANEOUS_USE_KHR` is a capability reported by the device when queried for `CL_DEVICE_COMMAND_BUFFER_CAPABILITIES_KHR`. If the device does not support that capability, then a test trying to verify simultaneous use functionality should skip, as negative testing isn't part of this test plan. However, if a device does support `CL_COMMAND_BUFFER_CAPABILITY_SIMULTANEOUS_USE_KHR`, then to use this functionality a test must pass `CL_COMMAND_BUFFER_SIMULTANEOUS_USE_KHR` when creating a command-buffer to enable that property on the command-buffer itself. So the behaviour is indeed a property of a command-buffer, but that behaviour can only be set if the device reports support for it.


> 3.3.1: "Test that an event returned by a command-buffer enqueue can be waited on an enqueue of the same command-buffer" - this does require simultaneous use.
I suggest you add a test where you query CL_DEVICE_COMMAND_BUFFER_REQUIRED_QUEUE_PROPERTIES_KHR and then create a queue with such properties for testing command buffers (since Mobieye's implementation requires CL_QUEUE_NO_SYNC_OPERATIONS_INTEL, without this all the tests would be skipped).

command-buffer objects are not guaranteed to work with any queue across all implementations. The `CL_DEVICE_COMMAND_BUFFER_REQUIRED_QUEUE_PROPERTIES_KHR` query returns any properties a command-queue must have in order to be compatible with a command-buffer, otherwise `clCreateCommandBufferKHR` will fail. Currently, we just try to skip tests if the properties with a queue don't match - https://github.com/KhronosGroup/OpenCL-CTS/blob/main/test_conformance/extensions/cl_khr_command_buffer/basic_command_buffer.cpp#L48 but the suggestion that tests should be more proactive in trying to create a new queue with the required properties where possible makes sense. An example of an existing test that creates a new queue, but could be smarter about it - https://github.com/KhronosGroup/OpenCL-CTS/blob/main/test_conformance/extensions/cl_khr_command_buffer/basic_command_buffer.cpp#L505-L506 

I think designing this well is probably the most interesting engineering challenge of this whole task, compared to writing the individual test bodies.

**Comment by aharon-abramson** [comment]
"I think designing this well is probably the most interesting engineering challenge of this whole task, compared to writing the individual test bodies." - I agree :)

**Comment by mhajder-mobica** [comment]
Hello, above description states in point 2.0:

> Test case for out-of-order command-buffer:

Please correct me if I am wrong, I would say this one is already covered by OutOfOrderTest, right ?

**Comment by EwanC** [comment]
> Hello, above description states in point 2.0:
> 
> > Test case for out-of-order command-buffer:
> 
> Please correct me if I am wrong, I would say this one is already covered by OutOfOrderTest, right ?

Yup, you are correct. One of the cases in "3.3.1 Returned events" I think is also covered by `UserEventTest`. So i've updated the comment to reflect that.
